node/core/chain-selection/src/lib.rs

```diff
@@ -0,0 +1,757 @@
+type Timestamp = u64;
+
+#[derive(Debug, Clone)]
+enum Approval {
+	// Approved
+	Approved,
+	// Unapproved but not stagnant
+	Unapproved,
+	// Unapproved and stagnant.
+	Stagnant,
+}
+
+impl Approval {
+	fn is_stagnant(&self) -> bool {
+		match *self {
+			Approval::Stagnant => true,
+			_ => false,
+		}
+	}
+}
+
+#[derive(Debug, Clone)]
+struct ViabilityCriteria {
+	// Whether this block has been explicitly reverted by one of its descendants.
+	explicitly_reverted: bool,
+	// `None` means approved. `Some` means unapproved.
+	approval: Approval,
+	earliest_unviable_ancestor: Option<Hash>,
+}
+
+impl ViabilityCriteria {
+	fn is_viable(&self) -> bool {
+		self.is_parent_viable() && self.is_partially_viable()
+	}
+
+	// Whether the current block is partially viable.
+	// That is, whether the current block is neither reverted nor stagnant.
+	fn is_partially_viable(&self) -> bool {
+		!self.explicitly_reverted && !self.approval.is_stagnant()
+	}
+
+	// Whether the parent is viable.
+	fn is_parent_viable(&self) -> bool {
+		self.earliest_unviable_ancestor.is_none()
+	}
+}
+
+#[derive(Debug, Clone)]
+struct LeafEntry {
+	weight: Weight,
+	block_hash: Hash,
+}
+
+#[derive(Debug, Clone)]
+struct LeafEntrySet {
+	inner: Vec<LeafEntry>
+}
+
+impl LeafEntrySet {
+	fn contains(&self, hash: &Hash) -> bool {
+		self.inner.iter().position(|e| &e.block_hash == hash).is_some()
+	}
+
+	fn remove(&mut self, hash: &Hash) -> bool {
+		match self.inner.iter().position(|e| &e.block_hash == hash) {
+			None => false,
+			Some(i) => {
+				self.inner.remove(i);
+				true
+			}
+		}
+	}
+
+	fn insert(&mut self, new: LeafEntry) {
+		match self.inner.iter().position(|e| e.weight < new.weight) {
+			None => self.inner.push(new),
+			Some(i) => if self.inner[i].block_hash != new.block_hash {
+				self.inner.insert(i, new);
+			}
+		}
+	}
+}
+
+#[derive(Debug, Clone)]
+struct BlockEntry {
+	block_hash: Hash,
+	parent_hash: Hash,
+	children: Vec<Hash>,
+	viability: ViabilityCriteria,
+	weight: Weight,
+}
+
+impl BlockEntry {
+	fn non_viable_ancestor_for_child(&self) -> Option<Hash> {
+		if self.viability.is_viable() {
+			None
+		} else {
+			self.viability.earliest_unviable_ancestor.or(Some(self.block_hash))
+		}
+	}
+}
+
+#[derive(Debug, thiserror::Error)]
+#[allow(missing_docs)]
+pub enum Error {
+	#[error(transparent)]
+	ChainApi(#[from] ChainApiError),
+
+	#[error(transparent)]
+	Io(#[from] std::io::Error),
+
+	#[error(transparent)]
+	Oneshot(#[from] oneshot::Canceled),
+
+	#[error(transparent)]
+	Subsystem(#[from] SubsystemError),
+
+	#[error(transparent)]
+	Codec(#[from] CodecError),
+}
+
+impl Error {
+	fn trace(&self) {
+		match self {
+			// don't spam the log with spurious errors
+			Self::Oneshot(_) => tracing::debug!(target: LOG_TARGET, err = ?self),
+			// it's worth reporting otherwise
+			_ => tracing::warn!(target: LOG_TARGET, err = ?self),
+		}
+	}
+}
+
+fn timestamp_now() -> Timestamp {
+	match SystemTime::now().duration_since(UNIX_EPOCH) {
```

Comment:
For our use-case of detecting stagnation in 90 seconds we shouldn't need to worry about whether the timestamp is perfectly accurate or even monotonic.

Reply:
My concern was mostly when non-monotonic comes into play, and the time advances unexpectedly. Then the timeout detection triggers earlier (technically could even instantly). If this is not something to consider (i.e. summer to wintertime change with system clock change) then leave as is

Reply:
I see what you mean, that the timeout could trigger instantly. I'm not very familiar with the internal workings of the clock but I imagine if a clock is slipping by 90 seconds something is seriously wrong. I would expect slippage of no more than a few seconds and typically less, and due to network asynchrony validators are expected to have slightly different views of when blocks become stagnant anyway.

Reply:
I can add a comment to address the concern about non-monotonicity, if that would help.

Reply:
Usually this is a more windows centric issue, where the OS happily messes with the system time.

A simple one liner comment would fine, since we don't currently target windows as a supported OS.